nvflare/app_opt/lightning/fl_callbacks.py

```diff
@@ -0,0 +1,99 @@
+from enum import IntEnum
+
+import pytorch_lightning as pl
+from pytorch_lightning.trainer.states import TrainerFn
+
+import nvflare.client as flare
+from nvflare.app_common.abstract.fl_model import FLModel
+from nvflare.client.config import ConfigKey
+from nvflare.client.constants import ModelExchangeFormat
+
+
+class SendTrigger(IntEnum):
+    AFTER_TRAIN_AND_TEST = 1
+    AFTER_TRAIN = 2
+    AFTER_TEST = 3
+
+
+class FLCallback(pl.callbacks.Callback):
+    def __init__(self, send_trigger: SendTrigger = SendTrigger.AFTER_TRAIN_AND_TEST):
+        super(FLCallback, self).__init__()
+        flare.init(
+            config={
+                ConfigKey.EXCHANGE_PATH: "./",
+                ConfigKey.EXCHANGE_FORMAT: ModelExchangeFormat.PYTORCH,
+                ConfigKey.TRANSFER_TYPE: "FULL",
+            }
+        )
+        self.send_mode = send_trigger
+        self.input_fl_model = None
+        self.output_fl_model = None
+        self.model_delivered = False
+
+    def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: str) -> None:
+        loop = trainer.test_loop
+        loop.run = test_loop_run_decorator(loop, self)
+
+    def on_fit_start(self, trainer, pl_module):
+        self._receive_update_model(pl_module)
+
+    def on_train_end(self, trainer, pl_module):
+        if self.output_fl_model:
+            self.output_fl_model.params = pl_module.cpu().state_dict()
+        else:
+            self.output_fl_model = flare.FLModel(params=pl_module.cpu().state_dict())
+
+    def on_test_start(self, trainer, pl_module):
+        if pl_module:
+            self._receive_update_model(pl_module)
```

Comment:
Note that this will change the behavior of the original "test" method.
But since we are already changing the behavior of "fit" method.
Maybe it does not matter.
If users choose to attach/use our hook in their Trainer, their fit/train/test method won't be the same as before.